localization/ndt_scan_matcher/src/map_update_module.cpp

```diff
@@ -69,11 +88,51 @@ void MapUpdateModule::update_map(const geometry_msgs::msg::Point & position)
     }
     status = result.wait_for(std::chrono::seconds(1));
   }
-  update_ndt(result.get()->new_pointcloud_with_ids, result.get()->ids_to_remove);
+  update_ndt(ndt, result.get()->new_pointcloud_with_ids, result.get()->ids_to_remove);
+}
+
+void MapUpdateModule::update_map(const geometry_msgs::msg::Point & position)
+{
+  // If the current position is super far from the previous loading position,
+  // lock and rebuild ndt_ptr_
+  if (need_rebuild_) {
+    ndt_ptr_mutex_->lock();
+    auto param = ndt_ptr_->getParams();
+
+    ndt_ptr_.reset(new NormalDistributionsTransform);
+
+    ndt_ptr_->setParams(param);
+
+    prefetch_map(position, ndt_ptr_);
+    ndt_ptr_mutex_->unlock();
+    need_rebuild_ = false;
+  } else {
+    // Load map to the secondary_ndt_ptr, which does not require a mutex lock
+    // Since the update of the secondary ndt ptr and the NDT align (done on
+    // the main ndt_ptr_) overlap, the latency of updating/alignment reduces partly.
+    // If the updating is done the main ndt_ptr_, either the update or the NDT
+    // align will be blocked by the other.
+    prefetch_map(position, secondary_ndt_ptr_);
+
+    ndt_ptr_mutex_->lock();
+    auto input_source = ndt_ptr_->getInputSource();
+    ndt_ptr_ = secondary_ndt_ptr_;
+    ndt_ptr_->setInputSource(input_source);
+    ndt_ptr_mutex_->unlock();
+  }
+
+  secondary_ndt_ptr_.reset(new NormalDistributionsTransform);
+  *secondary_ndt_ptr_ = *ndt_ptr_;
+
+  // Memorize the position of the last update
   last_update_position_ = position;
+
+  // Publish the new ndt maps
+  publish_partial_pcd_map();
 }
 
 void MapUpdateModule::update_ndt(
+  NdtPtrType & ndt,
```

Comment:
This may just be my stupidity, but I feel that using pointer references makes the code more complex and difficult to understand.
How about making it a function that receives a reference to NormalDistributionsTransform?

Reply:
@SakodaShintaro Done!